//
//  APIManager.swift
//  DemoApplication
//
//  Created by Sheetal on 20/07/20.
//  Copyright Â© 2020 Sheetal.com. All rights reserved.
//

import UIKit
import SwiftyJSON
enum ApiResult {
    case success(JSON)
    case failure(RequestError)
}
enum HTTPMethod: String {
    case options = "OPTIONS"
    case get     = "GET"
    case head    = "HEAD"
    case post    = "POST"
    case put     = "PUT"
    case patch   = "PATCH"
    case delete  = "DELETE"
    case trace   = "TRACE"
    case connect = "CONNECT"
}
enum RequestError: Error {
    case unknownError
    case connectionError
    case authorizationError(JSON)
    case invalidRequest
    case notFound
    case invalidResponse
    case serverError
    case serverUnavailable
}
protocol APIServiceProtocol {
    func requestData(url:String,completion: @escaping (ApiResult)->Void)
}

class APIManager: APIServiceProtocol {
    let baseUrl = "https://dl.dropboxusercontent.com/"
    typealias parameters = [String:Any]
    func requestData(url: String, completion: @escaping (ApiResult) -> Void) {
        var request = URLRequest(url: URL(string: baseUrl+url)!)
        request.httpMethod = "GET"
        URLSession.shared.dataTask(with: request) { (data, response, error) in
            if let error = error {
                print(error)
                completion(ApiResult.failure(.connectionError))
            }else if let data = data ,let responseCode = response as? HTTPURLResponse {
                do {
                    let utf8Data = String(decoding: data, as: UTF8.self).data(using: .utf8)
                    let responseJson = try JSON(data: utf8Data!)
                    print("responseCode : \(responseCode.statusCode)")
                    print("responseJSON : \(responseJson)")
                    switch responseCode.statusCode {
                    case 200:
                        completion(ApiResult.success(responseJson))
                    case 400...499:
                        completion(ApiResult.failure(.authorizationError(responseJson)))
                    case 500...599:
                        completion(ApiResult.failure(.serverError))
                    default:
                        completion(ApiResult.failure(.unknownError))
                        break
                    }
                }
                catch let parseJSONError {
                    completion(ApiResult.failure(.unknownError))
                    print("error on parsing request to JSON : \(parseJSONError)")
                }
            }
        }.resume()
    }
}
